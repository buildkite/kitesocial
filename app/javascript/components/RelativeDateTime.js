import * as React from 'react';
import PropTypes from 'prop-types';

import { DateTime } from 'luxon';

const TARGET_UNITS = ['years', 'months', 'weeks', 'days', 'hours', 'minutes', 'seconds'];
const FORMATTER = new Intl.RelativeTimeFormat();

const formatDate = (date) => {
  const elapsed = date.diffNow();

  const unit = TARGET_UNITS.find((unit) => Math.abs(elapsed.as(unit)) >= 1);
  if (!unit) {
    return 'just now';
  }

  const unitValue = Math.round(elapsed.as(unit));

  return FORMATTER.format(unitValue, unit);
};

const RelativeDateTime = ({ value }) => {
  const date = DateTime.fromISO(value);

  return (
    <time
      title={date.toLocaleString(DateTime.DATETIME_SHORT)}
      dateTime={value}
    >
      {formatDate(date)}
    </time>
  );
};

RelativeDateTime.propTypes = {
  value: PropTypes.string.isRequired
};

export default RelativeDateTime;
