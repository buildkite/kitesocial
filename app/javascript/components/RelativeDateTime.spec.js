/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import RelativeDateTime from './RelativeDateTime';

beforeAll(() => {
  jest.useFakeTimers('modern');
  jest.setSystemTime(new Date('2021-04-22T01:08:10Z').getTime())
});

afterAll(() => {
  jest.useRealTimers();
});

describe('rendering a relative date', () => {
  it('handles values which are less than a second different from the current time', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-22T01:08:10Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values measured in seconds', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-22T01:08:01Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values measured in minutes', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-22T01:02:01Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values measured in hours', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-21T22:02:01Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values measured in days', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-18T22:02:01Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values measured in weeks', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-01T22:02:01Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values measured in months', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-02-01T22:02:01Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles values greater than a year old', () => {
    const component = renderer.create(
      <RelativeDateTime value="2020-04-22T00:01:25Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles negative values measured in seconds', () => {
    const component = renderer.create(
      <RelativeDateTime value="2021-04-22T01:08:45Z" />,
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
