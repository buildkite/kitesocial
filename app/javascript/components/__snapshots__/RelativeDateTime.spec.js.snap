// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering a relative date handles negative values measured in seconds 1`] = `
<time
  dateTime="2021-04-22T01:08:45Z"
  title="2021-04-21, 6:08 p.m."
>
  in 35 seconds
</time>
`;

exports[`rendering a relative date handles values greater than a year old 1`] = `
<time
  dateTime="2020-04-22T00:01:25Z"
  title="2020-04-21, 5:01 p.m."
>
  1 year ago
</time>
`;

exports[`rendering a relative date handles values measured in days 1`] = `
<time
  dateTime="2021-04-18T22:02:01Z"
  title="2021-04-18, 3:02 p.m."
>
  3 days ago
</time>
`;

exports[`rendering a relative date handles values measured in hours 1`] = `
<time
  dateTime="2021-04-21T22:02:01Z"
  title="2021-04-21, 3:02 p.m."
>
  3 hours ago
</time>
`;

exports[`rendering a relative date handles values measured in minutes 1`] = `
<time
  dateTime="2021-04-22T01:02:01Z"
  title="2021-04-21, 6:02 p.m."
>
  6 minutes ago
</time>
`;

exports[`rendering a relative date handles values measured in months 1`] = `
<time
  dateTime="2021-02-01T22:02:01Z"
  title="2021-02-01, 2:02 p.m."
>
  3 months ago
</time>
`;

exports[`rendering a relative date handles values measured in seconds 1`] = `
<time
  dateTime="2021-04-22T01:08:01Z"
  title="2021-04-21, 6:08 p.m."
>
  9 seconds ago
</time>
`;

exports[`rendering a relative date handles values measured in weeks 1`] = `
<time
  dateTime="2021-04-01T22:02:01Z"
  title="2021-04-01, 3:02 p.m."
>
  3 weeks ago
</time>
`;

exports[`rendering a relative date handles values which are less than a second different from the current time 1`] = `
<time
  dateTime="2021-04-22T01:08:10Z"
  title="2021-04-21, 6:08 p.m."
>
  just now
</time>
`;
